src/common/upload.ts

```diff
@@ -118,36 +125,80 @@ export abstract class TurboAuthenticatedBaseUploadService
     dataItemOpts,
   }: TurboFileFactory &
     TurboAbortSignal): Promise<TurboUploadDataItemResponse> {
-    const { dataItemStreamFactory, dataItemSizeFactory } =
-      await this.signer.signDataItem({
-        fileStreamFactory,
-        fileSizeFactory,
-        dataItemOpts,
-      });
-    const signedDataItem = dataItemStreamFactory();
-    this.logger.debug('Uploading signed data item...');
-    // TODO: add p-limit constraint or replace with separate upload class
+    let retries = 0;
+    const maxRetries = this.retryConfig.retries ?? 3;
+    const retryDelay =
+      this.retryConfig.retryDelay ??
+      ((retryNumber: number) => retryNumber * 1000);
+    let lastError = 'Unknown Error'; // Store the last error for throwing
+    let lastStatusCode: number | undefined; // Store the last status code for throwing
+
+    while (retries < maxRetries) {
+      if (signal?.aborted) {
+        throw new CanceledError();
+      }
 
-    const headers = {
-      'content-type': 'application/octet-stream',
-      'content-length': `${dataItemSizeFactory()}`,
-    };
-    if (dataItemOpts !== undefined && dataItemOpts.paidBy !== undefined) {
-      const paidBy = Array.isArray(dataItemOpts.paidBy)
-        ? dataItemOpts.paidBy
-        : [dataItemOpts.paidBy];
+      const { dataItemStreamFactory, dataItemSizeFactory } =
+        await this.signer.signDataItem({
```

Comment:
What's the UX for retries as it pertains to signing? Might the user be prompted for their sig again?

Reply:
unfortunately yes, because in the Node flow (and ideally the web flow as well, we have a TODO there) we're consuming the stream in order to start the signing process and then sending the bytes through to the request. so on network failure the whole stream needs to be recreated because bytes have been consumed

we could iterate later provide an option here to use an ephemeral signer, that would ensure the user signs once no matter the retries. that option would also be more useful in bulk uploads -- theres a ticket on the board for that

Reply:
I almost wonder if we need another set of APIs that lets you compute the signature once and then if you have it, use it in combination with a file stream to give you a data item stream that does not require signer invocation again. Something we can iterate on.